File: go-sdk/property_repository.go
package cms

import "sync"

// PropertyRepository stores property values as UTF-8 text, mirroring the Java SDK.
type PropertyRepository interface {
	// GetByKey returns nil if the key is absent (including when the value was never set).
	GetByKey(key string) *string

	// Store sets the value for key when newValue is non-nil, or removes the key when newValue is nil.
	// It returns the previous value, or nil if the key was absent.
	Store(key string, newValue *string) *string
}

type InMemoryPropertyRepository struct {
	mu         sync.RWMutex
	properties map[string]string
}

func NewInMemoryPropertyRepository() *InMemoryPropertyRepository {
	return &InMemoryPropertyRepository{
		properties: make(map[string]string),
	}
}

func (r *InMemoryPropertyRepository) GetByKey(key string) *string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	v, ok := r.properties[key]
	if !ok {
		return nil
	}
	return &v
}

func (r *InMemoryPropertyRepository) Store(key string, newValue *string) *string {
	r.mu.Lock()
	defer r.mu.Unlock()
	old, hadOld := r.properties[key]
	var oldPtr *string
	if hadOld {
		o := old
		oldPtr = &o
	}
	if newValue == nil {
		delete(r.properties, key)
	} else {
		r.properties[key] = *newValue
	}
	return oldPtr
}

File: python-sdk/cms/property_repository.py
"""Thread-safe in-memory property storage."""

from __future__ import annotations

import threading
from abc import ABC, abstractmethod
from typing import Optional


class PropertyRepository(ABC):
    """Interface for key/value property storage."""

    @abstractmethod
    def get_by_key(self, key: str) -> Optional[str]:
        """Return the stored value for *key*, or ``None`` if absent."""

    @abstractmethod
    def store(self, key: str, new_value: Optional[str]) -> Optional[str]:
        """Store *new_value* under *key*.

        If *new_value* is ``None`` the key is removed.
        Returns the previous value (or ``None`` if the key was absent).
        """


class InMemoryPropertyRepository(PropertyRepository):
    """Concurrency-safe in-memory implementation backed by a plain ``dict``."""

    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._properties: dict[str, str] = {}

    def get_by_key(self, key: str) -> Optional[str]:
        with self._lock:
            return self._properties.get(key)

    def store(self, key: str, new_value: Optional[str]) -> Optional[str]:
        with self._lock:
            old = self._properties.get(key)
            if new_value is None:
                self._properties.pop(key, None)
            else:
                self._properties[key] = new_value
            return old
